tools.go

```diff
@@ -0,0 +1,43 @@
+package resolver
+
+import (
+	"fmt"
+	"strings"
+
+	"darvaza.org/core"
+	"github.com/miekg/dns"
+	"golang.org/x/net/idna"
+)
+
+func sanitiseNetwork(network string) (string, error) {
+	s := strings.ToLower(network)
+	switch s {
+	case "":
+		return "ip", nil
+	case "ip", "ip4", "ip6":
+		return s, nil
+	default:
+		return "", fmt.Errorf("%q: invalid network", network)
+	}
+}
+
+func sanitiseHost(host string) (string, error) {
+	s, err := idna.Display.ToASCII(host)
+	if err != nil {
+		return "", core.Wrapf(err, "%q: invalid host", host)
+	}
+	return s, nil
+}
+
+// ForEachRR calls a function for each answer of the specified type.
+func ForEachRR[T any](msg *dns.Msg, fn func(v T)) {
```

Comment:
if this is a `RR` then why do you apply the `fn` only to the `Answer` section?
in miekg the Msg contains 3 slices of `RR` `Answer`, `Ns`, and `Extra` with `len` not necessarily bigger than zero. Since iterating over a zero length slice is not an error, I would expect a function named `ForEachRR()` to apply the `fn` to all three slices, or be called `ForEachAnswer()`.

Reply:
Thanks, I'll amend it